### PR TITLE
global sort: add 2-level iterator for merge property and range splitter

### PR DESCRIPTION
**Pull request by lance6716**
### What problem does this PR solve?

Issue Number: close https://github.com/pingcap/tidb/issues/49717

Problem Summary:

### What changed and how does it work?

- For `MultipleFilesStat`, it's maintained every 500 files, and `MultipleFilesStat.MaxOverlappingNum` is the number of the key range maximum overlaping of the 500 files. In this PR, `MultipleFilesStat.Filenames` is also sorted by start key of each file, so that we only need to keep number of `MultipleFilesStat.MaxOverlappingNum` files open from left to right, and get the minimum key among the opened files.
- We introduced a `limitSizeMergeIter`, which can have reader openers, their weights, and maximum weight limit as arguments. `limitSizeMergeIter` will automatically open readers in given order and keep readers' weight sum under the limit.
- For `RangeSplitter`, now it takes `[]MultipleFilesStat` as arguments. It has two levels of `limitSizeMergeIter`. The upper level is a whole `MultipleFilesStat` with weight `MultipleFilesStat.MaxOverlappingNum`, and the limit for this level is `MergeSortOverlapThreshold`. The second level is intra-`MultipleFilesStat` which means the 500 files, with weight `1`, and the limit is `MultipleFilesStat.MaxOverlappingNum`. In this way `RangeSplitter` can still output ordered key but won't open all files.

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

**Review by D3Hunter — COMMENTED** [review]
12/13

File: br/pkg/lightning/backend/external/iter.go [hunk 1]
```diff
@@ -292,23 +313,121 @@ func (i *mergeIter[T, R]) next() bool {
...
+	i := &limitSizeMergeIter[T, R]{
+		mergeIter:     iter,
+		readerOpeners: readerOpeners,
+		weights:       weights,
+		nextReaderIdx: end,
+		limit:         limit,
+	}
+	for j := 0; j < end; j++ {
+		i.weightSum += weights[j]
+	}
```
Comment (D3Hunter) [inline br/pkg/lightning/backend/external/iter.go:385]:
```suggestion
	i.weightSum = cur
```

File: br/pkg/lightning/backend/external/iter.go [hunk 2]
```diff
@@ -292,23 +313,121 @@ func (i *mergeIter[T, R]) next() bool {
...
+}
+
+// limitSizeMergeIter acts like a mergeIter, except that each reader has a weight
+// and it will try to open more readers with the total weight doesn't exceed the
+// limit.
+//
+// Because it's like a mergeIter it's expected to iterate in ascending order,
+// caller should set a proper limit to do not block opening new readers
+// containing the next minimum element.
+type limitSizeMergeIter[T heapElem, R sortedReader[T]] struct {
```
Comment (D3Hunter) [inline br/pkg/lightning/backend/external/iter.go:343]: if the files are ordered by firstKey, we can open those files when current iterated key >= firstKey, so no need to calculate weight.

it's error-prone to let caller make sure the limit is enough and might cause the iterated key might not in ascending order.

Comment (lance6716) [reply br/pkg/lightning/backend/external/iter.go:343]: We need to open the file to know the first key (or record first key in another place). In current design, we don't record first key for intra-`MultipleFilesStat`. Another point is it's more cheap to saving a number representing the maximum overlapping level than saving every start key.

> make sure the limit is enough and might cause the iterated key might not in ascending order.

the caller can use `mergeIter` directly, like `MergeKVIter`. Currently for this problem I can only figure out `limitSizeMergeIter` with this prerequisite.

Comment (lance6716) [reply br/pkg/lightning/backend/external/iter.go:343]: And in fact in my first design half a year ago I suggest to record `MinKey`, `MaxKey` for every file and save it in a `MetadataStorage` interface (a table in downstream database should be OK), so these actions are easy to implement 😂 Current global sort system component design is complicated and I think they are over-optimized.

**Review by D3Hunter — APPROVED** [review]

**Review by Benjamin2037 — APPROVED** [review]
LGTM

**Pull request merged**